### PR TITLE
Adjust to variable result grouping

### LINKED ISSUE #25: feat: sankey plot only depends on inputs generated with `pacta.multi.loanbook.analysis`
currently the sankey plot uses company level alignment metric as input, combined with the matched prioritized loan books for exposure.

we want to have all plots depend only on outputs generated with `pacta.multi.loanbook.analysis`. This means that the sankey plot should be using the loan book aggregated alignment metric instead. 

One implication of this is that it will not be possible to display company level exposures in the sankey plot as a standard feature, but this has turned out not to be very useful anyway. instead, we may review later on, if the open grouping variable that is to be added for P4S analysis should be allowed in the sankey plot, which would be consistent with using `pacta.multi.loanbook.analysis` outputs only

AB#10526

### LINKED ISSUE #27: feat: ensure all plots can handle the variable filtering/grouping functionality from `pacta.multi.loanbook.analysis`
When we add variable filtering/grouping to `pacta.multi.loanbook.analysis`, this will affect the output formats in that the data will have columns `id_loanbook` (to identify the source loan book) and `group_var` (to identify any additional variable the calculations may be grouped by), instead of only `group_id` (which stood in for the loan book identifier).

The expectations across the `prep_*` and `plot_*` functions need to be adjusted, and where necessary an option to display the `group_var` should be added.

relates to
- https://dev.azure.com/RMI-PACTA/2DegreesInvesting/_workitems/edit/10322
- https://dev.azure.com/RMI-PACTA/2DegreesInvesting/_workitems/edit/10323

AB#10497

### PR DESCRIPTION
**Pull request by jacobvjk**
depends on https://github.com/RMI-PACTA/pacta.multi.loanbook.analysis/pull/34
closes #27 
closes #25 by coincidence

- adapts the `pacta.multi.loanbook.plot` package to changes made in `pacta.multi.loanbook.analysis`, which now allows aggregating results by any user defined variable
- all `prep_*()` and `plot_*()` functions gain flexibility in processing grouping variables
  - Where they previously expected a variable `"group_id"`, the variable can now be specified to any name using the `group_var` argument
  - Omitting the argument aggregates across all input loan books, allowing for a meta view
- coincidentally, the linked changes in `pacta.multi.loanbook.analysis` remove the need to use the `matched_prioritized` data set as an input to the sankey plot. this means that `pacta.multi.loanbook.plot` now solely uses data objects created with `pacta.multi.loanbook.analysis`
- aligns on using `group_var` in function signature, changing `category` to `group_var` in `prep_scatter_alignment_exposure()` and `plot_scatter_alignment_exposure()`
- references to `"bank"` in scatter plots are changed to `"group"` and `data_level` can now be `"company"` or `"group_var"`, not `"bank"`

### CODE REVIEW

**Comment by jacobvjk** [comment]
@jdhoffa @MonikaFu I am still keeping this in draft mode, as we probably want to agree on the linked PR first. But it may make sense to take a brief look at this PR now as well. Explanation is above, but just a quick tl;dr:

the aggregate metric needs to be flexible for analysis, allowing slicing and dicing of the loan books by user defined variables (e.g. by banks, but possibly also by membership in alliances, loan structures, etc). The prep and plot functions should cover this in a basic way but could probably use some prettification. open to ideas and suggestions here

**Comment by jdhoffa** [comment]
Ah this might be a bit of an anti-pattern. 

Specifically for the `prep_` and `plot_` functions, I would tend to prefer that they always expect a single unit of input to `prep` or `plot` (so as to not overcomplicate them). And if iterating over groups is desired, I would imagine doing that either in:
* the`workflow.*` 
* some wrapper function
* a vignette

is probably desired? 

I don't think hard-baking the grouping functionality into the plotting functions is a good idea...

**Comment by jacobvjk** [comment]
> Ah this might be a bit of an anti-pattern.
> 
> Specifically for the `prep_` and `plot_` functions, I would tend to prefer that they always expect a single unit of input to `prep` or `plot` (so as to not overcomplicate them). And if iterating over groups is desired, I would imagine doing that either in:
> 
> * the`workflow.*`
> * some wrapper function
> * a vignette
> 
> is probably desired?
> 
> I don't think hard-baking the grouping functionality into the plotting functions is a good idea...

It's definitely not pretty at this point and it may be worth reviewing which of the plots realisitically need to show any group information. But I think in some cases we explicitly want the group info inside the plot. See e.g. https://github.com/RMI-PACTA/workflow.aggregate.loanbooks/pull/95

**Comment by jdhoffa** [comment]
Ok got it, then totally fair enough!

File: R/plot_sankey.R
Comment (jdhoffa) [inline R/plot_sankey.R:5]: (very) NIT: but the name `by_group` seems like it should take a boolean?

Comment (jacobvjk) [reply R/plot_sankey.R:5]: any other ideas for naming this? `group` does not seem much better.. `.by` would seem odd here?

Comment (jdhoffa) [reply R/plot_sankey.R:5]: `sankey_group`?

Comment (jacobvjk) [reply R/plot_sankey.R:5]: it is being used for all the different plots... so it would have to be `<plot>_group` then.. not sure

File: R/plot_timeline.R
Comment (MonikaFu) [inline R/plot_timeline.R:5]: I might not entirely understand the logic here but my impression is that if there is a `by_group` column present in the input data this is used for grouping the data in plots, otherwise all data is used and not grouped in any way? In that case, do we need to have it both as a requirement for the data and as a separate argument?

File: R/prep_sankey.R
Comment (MonikaFu) [inline R/prep_sankey.R:4]: same comment as above

File: R/prep_scatter.R
Comment (MonikaFu) [inline R/prep_scatter.R:17]: Here 'Bank' should probably be changed to something else?

Comment (jacobvjk) [reply R/prep_scatter.R]: indeed

File: R/plot_timeline.R
Comment (jacobvjk) [reply R/plot_timeline.R:5]: not sure I fully understand your suggestion. Do you mean it should be sufficient to use `by_group` in the `prep_*() `functions and then just use whatever "additional" variables are available after `prep_*()` for grouping in `plot_*()`, thereby removing the need to explicitly set `by_group` in any of the `plot_*()` functions?

Comment (MonikaFu) [reply R/plot_timeline.R:5]: I meant that I don't understand why we require both `by_group` parameter and then also we require that a column called `by_group` is present. Now I understand that `by_group` is actually varying. We should make it clearer in the documentation that by_group is not a column name but a variable. I think you did it in one of the functions.

File: R/plot_sankey.R
Comment (jacobvjk) [reply R/plot_sankey.R:5]: naming things has go to be the worst exercise haha. I think `group_var` is good. Weirdly enough that is what I wanted to use in `pacta.multi.loanbook.analysis` before opting for `.by`. But since it does something slightly different there, I would be find with using `group_var` here

File: R/plot_timeline.R
Comment (jacobvjk) [reply R/plot_timeline.R:5]: I see, makes sense

File: R/plot_sankey.R
Comment (jdhoffa) [reply R/plot_sankey.R:5]: There are only two hard things in Computer Science: cache invalidation and naming things.

-- Phil Karlton

File: R/plot_timeline.R
Comment (jacobvjk) [reply R/plot_timeline.R]: updated

File: R/prep_sankey.R
Comment (jacobvjk) [reply R/prep_sankey.R]: updated

**Review by jdhoffa — APPROVED** [review]
LGTM!
though I leave it to @MonikaFu to give the final approval given she is maintainer :-)

**Review by MonikaFu — APPROVED** [review]
LGTM. Tested with https://github.com/RMI-PACTA/workflow.aggregate.loanbooks/pull/95 and using different `group_id` variables. We might consider more elegant implementation indeed in the long run but given the uncertainty around the plots at the moment I'd say this is good enough.

**Pull request merged**
